lib/TH/generic/THStorage.c

```diff
@@ -107,9 +107,9 @@ void THStorage_(free)(THStorage *storage)
   if(!storage)
     return;
 
-  if((storage->flag & TH_STORAGE_REFCOUNTED) && (storage->refcount > 0))
+  if((storage->flag & TH_STORAGE_REFCOUNTED) && (THAtomicGet(&storage->refcount) > 0))
   {
```

Comment:
>When I started using atomics in djvu, I decided not to use atomic gets and
sets on aligned elementary types (1,2,4,8 bytes longs things) because I do not
believe that reasonable hardware would break such aligned accesses into
multiple parts. I only use them for basic read-modify-write operations such as
increment, decrement, compare-and-swap, test-and-set, etc.

FYI, now that C11/C++11 has an explicit memory model (e.g., http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf, section 5.1.2.4), this behavior is very dangerous.

The issue is not just what the hardware does, but what the *compiler* is also allowed to do.

There's no guarantee that without a memory fence, a write that one thread makes must be visible at all to any other thread, or that a set of writes will be seen in order by other threads. The compiler is completely free to do whatever it wants to with this write, as long as the current thread only sees its writes in order. In fact, the standard basically indicates that it could take an arbitrary long period of time for writes by one thread to be made visible to another thread without fences or atomics, if at all. A compiler has to make some assumption across function calls and has to issue its writes out to memory before making them (unless it is optimizing them together and has visibility into what is being called), at which point you depend upon the hardware level memory visibility rules.

Such behavior makes a lot of old patterns now dangerous (http://plv.mpi-sws.org/c11comp/popl15.pdf), though the behavior was somewhat undefined before. Java, for instance, always had a rule that writes to 8 byte values were potentially non-atomic and were prone to tearing, since the underlying hardware could provide no such guarantees.

Getting back to the issue:

>Isn't this AtomicGet unnecessary? The ref counting for Tensors below just jumps straight to the DecrementRef.
>i hesitated to access atomic variables only with atomic functions, as an habit. (in fact, my first version had atomic as a struct, to refrain the user touching directly the number in there). true, here does not do much... i put it only for consistency reasons. you think i should remove it?

You need an atomic read here, since you are writing to it atomically, and there's no memory fence for any normal read implied (an atomic write provides no general fence guarantees for all writes, just for that since value, if it is implemented as a hardware-level CAS).

Rather I'd raise the question of why you are reading the value beforehand at all (checking that storage->refcount is > 0)? Why not just atomic decrement, and if zero, free?